SimTracker/TrackTriggerAssociation/plugins/StubAssociator.cc

```diff
@@ -0,0 +1,143 @@
+    putTokenSelection_ = produces<StubAssociation>(iConfig.getParameter<string>("BranchSelection"));
+    // book ES product
+    esGetToken_ = esConsumes<Setup, SetupRcd, Transition::BeginRun>();
+  }
+
+  void StubAssociator::beginRun(const Run& iRun, const EventSetup& iSetup) {
+    setup_ = &iSetup.getData(esGetToken_);
+  }
+
+  void StubAssociator::produce(Event& iEvent, const EventSetup& iSetup) {
+    // associate TTStubs with TrackingParticles
+    Handle<TTStubDetSetVec> handleTTStubDetSetVec;
+    iEvent.getByToken<TTStubDetSetVec>(getTokenTTStubDetSetVec_, handleTTStubDetSetVec);
+    Handle<TTClusterAssMap> handleTTClusterAssMap;
+    Handle<TTStubAssMap> handleTTStubAssMap;
+    if (useTTStubAssMap_)
+      iEvent.getByToken<TTStubAssMap>(getTokenTTStubAssMap_, handleTTStubAssMap);
+    else
+      iEvent.getByToken<TTClusterAssMap>(getTokenTTClusterAssMap_, handleTTClusterAssMap);
+    map<TPPtr, vector<TTStubRef>> mapTPPtrsTTStubRefs;
+    for (TTStubDetSetVec::const_iterator ttModule = handleTTStubDetSetVec->begin();
+        ttModule != handleTTStubDetSetVec->end();
+        ttModule++) {
+      for (TTStubDetSet::const_iterator ttStub = ttModule->begin(); ttStub != ttModule->end(); ttStub++) {
+        const TTStubRef ttStubRef = makeRefTo(handleTTStubDetSetVec, ttStub);
+        set<TPPtr> tpPtrs;
+        if (useTTStubAssMap_)
+          fill(tpPtrs, ttStubRef, handleTTStubAssMap);
+        else
+          fill(tpPtrs, ttStubRef, handleTTClusterAssMap);
+        for (const TPPtr& tpPtr : tpPtrs)
+          mapTPPtrsTTStubRefs[tpPtr].push_back(ttStubRef);
+      }
+    }
+    // associate reconstructable TrackingParticles with TTStubs
+    StubAssociation reconstructable(setup_);
+    StubAssociation selection(setup_);
+    for (const pair<TPPtr, vector<TTStubRef>>& p : mapTPPtrsTTStubRefs) {
+      if (!setup_->useForReconstructable(*p.first) || !setup_->reconstructable(p.second))
+        continue;
+      reconstructable.insert(p.first, p.second);
+      if (setup_->useForAlgEff(*p.first))
+        selection.insert(p.first, p.second);
+    }
+    iEvent.emplace(putTokenReconstructable_, move(reconstructable));
+    iEvent.emplace(putTokenSelection_, move(selection));
+  }
+
+  //
+  void StubAssociator::fill(set<TPPtr>& tpPtrs, const TTStubRef& ttStubRef, const Handle<TTClusterAssMap>& handle) const {
+    auto isNonnull = [](const TPPtr& tpPtr){ return tpPtr.isNonnull(); };
+    for (unsigned int iClus = 0; iClus < 2; iClus++) {
+      const vector<TPPtr>& assocPtrs = handle->findTrackingParticlePtrs(ttStubRef->clusterRef(iClus));
+      copy_if(assocPtrs.begin(), assocPtrs.end(), inserter(tpPtrs, tpPtrs.begin()), isNonnull);
+    }
+  }
+
+  //
+  void StubAssociator::fill(set<TPPtr>& tpPtrs, const TTStubRef& ttStubRef, const Handle<TTStubAssMap>& handle) const {
+    const TPPtr tpPtr = handle->findTrackingParticlePtr(ttStubRef);
```

Comment:
Use of TTStubAssMap.findTrackingParticlePtr() here is wrong. As explained in comment at https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/SimTracker/TrackTriggerAssociation/interface/TTStubAssociationMap.h#L70 , this function returns only the principle TP associated with each stub. However, the standard way we do truth-to-track association is to consider ALL TP associated to each stub. This is explained in the comments near https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/SimTracker/TrackTriggerAssociation/interface/TTTrackAssociationMap.h#L79 , and done in the standard track-to-truth associator at https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-12_0_0_pre4/SimTracker/TrackTriggerAssociation/plugins/TTTrackAssociator.cc#L53 . This explains the poor efficiencies reported above when UseTTStubAssMap=True.